### ISSUE TITLE
Top-down centered-instance pipeline

### ISSUE THREAD

**Issue by talmo** [talmolab/sleap-nn#3]
Add the core data pipeline for top-down centered instance models.

Use [`TopdownConfmapsPipeline`](https://github.com/talmolab/sleap/blob/c94c75517b34d6c14f4e3bd80695daab372fdf0a/sleap/nn/data/pipelines.py#L641) as reference.

Roughly:

1. Read labels (#1)
2. Augmentation (#2)
3. Find centroids/anchors (#7)
4. Crop instances (#13)
5. Apply image normalization (#9)
6. Generate confidence maps (#11)
7. Shuffling/Batching/prefetching

Some of these may need to occur in a different order than above or in the reference pipeline to work better with the augmentation backends and/or with the PyTorch Lightning data model.

**Notes:**
- Be careful when handling the number of instances. 1 instance = 1 example, but we may have a variable number of instances per frame. In the [reference implementation](https://github.com/talmolab/sleap/blob/c94c75517b34d6c14f4e3bd80695daab372fdf0a/sleap/nn/data/instance_cropping.py#L454-L456), we use `tf.data.Dataset.unbatch()` to go from frame-level examples to instance-level examples.
- Shuffling should be done at frame-level if possible
- Do image normalization as late as possible since copying `uint8` arrays to GPU will be 4x faster than `float32`
- Make sure to implement the same behavior of centroid/anchor detection as in the [reference pipeline](https://github.com/talmolab/sleap/blob/c94c75517b34d6c14f4e3bd80695daab372fdf0a/sleap/nn/data/instance_centroids.py#L55). Basically: prefer to use the specified anchor node, but if not visible use the midpoint of the bounding box.
- Potential performance optimization to explore (specific to this pipeline): applying augmentation to larger crops rather than full image. If we apply augmentation to the final crops, we'll get black patches at the corners if we rotate. If we apply augmentation to the full image, we don't get black edges, but it's much less efficient. The best solution would be to crop at `sqrt(2) * box_size` (or a bit more for rounding error) so that we never have black edges regardless of the rotation angle.

**Comment by talmo** [comment]
In light of https://github.com/pytorch/data/issues/1196, we should be mindful of using the existing patterns and functionality in core PyTorch's implementation of datapipes: https://github.com/pytorch/pytorch/tree/main/torch/utils/data/datapipes

Namely, we should consider adding decorators to register our DataPipes with the functional API (`@functional_datapipe("method_name")`), as well as making sure to reuse existing DataPipe blocks like shufflers, etc.

**Comment by talmo** [comment]
For high level pipeline builder, we would like an API that's really convenient and readable like:

```py
import sleap_io as sio
import sleap_nn as snn

labels = sio.load_slp("train.pkg.slp")
pipeline = snn.data.TopDownPipeline(labels=labels, anchor_node="thorax", crop_size=160, batch_size=4, rotation=180)
```

Maybe we'll want to use a `@classmethod` to construct it instead? Not clear if there is a usecase for it necessarily or if the `__init__` constructor will suffice.
